### PR TITLE
fix: add merge logic for exponential histogram when the temporality cumulative

### PR DESCRIPTION
**Pull request by xuan-cao-swi**
## Description
1. added merge logic for exponential histogram when the temporality cumulative 
2. added more test case cover merge logic and when scale/bucket size out of range

## Test
Test case are copied from python metrics sdk, and modified based on ruby metrics sdk data structure.

### CODE REVIEW

**Comment by akahn** [comment]
As an end-user what is the result of not having this fix? If we use exponential histograms, are they not accurate?

**Comment by xuan-cao-swi** [comment]
This pr is for adding merge logic for cumulative aggregation.
If you use delta aggregation, it should be accurate (no merge logic required since data will be dropped each time), but if you use cumulative, then it was expecting user's collector/collection side have the merge logic ; otherwise the result is unexpected.

File: metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb [hunk 1]
```diff
@@ -44,11 +45,25 @@ def initialize(
...
+            @previous_scale = {} # nil
+            # @start_time_unix_nano = {} #nil
           end
 
+          # when aggregation temporality is cumulative, merge and downscale will happen.
+          # rubocop:disable Metrics/MethodLength
           def collect(start_time, end_time, data_points)
             if @aggregation_temporality.delta?
               # Set timestamps and 'move' data point values to result.
+              # puts "data_points.inspect: #{data_points.inspect}"
```
Comment (kaylareopelle) [inline metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb]: do we need to keep the puts?

File: metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb [hunk 2]
```diff
@@ -44,11 +45,25 @@ def initialize(
...
+            # Previous state for cumulative aggregation
+            @previous_positive = {} # nil
+            @previous_negative = {} # nil
+            @previous_min = {} # Float::INFINITY
+            @previous_max = {} # -Float::INFINITY
+            @previous_sum = {} # 0
+            @previous_count = {} # 0
+            @previous_zero_count = {} # 0
+            @previous_scale = {} # nil
+            # @start_time_unix_nano = {} #nil
```
Comment (kaylareopelle) [inline metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb]: do we need this?

**Comment by akahn** [comment]
> expecting user's collector/collection side have the merge logic

We use Alloy as the collector—how do I know whether it is able to handle histogram merges? I'm trying to understand whether I need this patch in order to use exponential histograms with the Ruby SDK today.

**Comment by kaylareopelle** [comment]
@xuan-cao-swi  - Could you send me a link to the Python implementation you used? I'd like to compare the two.

File: metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_histogram/exponent_mapping.rb [hunk 3]
```diff
@@ -13,6 +13,9 @@ module ExponentialHistogram
           class ExponentMapping
             attr_reader :scale
 
+            MINIMAL_SCALE = -10
+            MAXIMAL_SCALE = 0
```
Comment (kaylareopelle) [inline metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_histogram/exponent_mapping.rb]: Where is the `MAXIMAL_SCALE` constant used?

File: metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_histogram/logarithm_mapping.rb [hunk 4]
```diff
@@ -13,6 +13,9 @@ module ExponentialHistogram
           class LogarithmMapping
             attr_reader :scale
 
+            MINIMAL_SCALE = 1
+            MAXIMAL_SCALE = 20
```
Comment (kaylareopelle) [inline metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_histogram/logarithm_mapping.rb]: Where are these constants used?

File: metrics_sdk/test/opentelemetry/sdk/metrics/aggregation/exponential_histogram/exponential_bucket_histogram_test.rb [hunk 5]
```diff
@@ -0,0 +1,1052 @@
...
+      assert_equal('Buckets max size 10000000 is larger than maximum max size 16384', error.message)
+
+      error = assert_raises(ArgumentError) do
+        OpenTelemetry::SDK::Metrics::Aggregation::ExponentialBucketHistogram.new(max_size: 0)
+      end
+      assert_equal('Buckets min size 0 is smaller than minimum min size 2', error.message)
+    end
+  end
+
+  # Integration tests moved from exponential_bucket_histogram_integration_test.rb
```
Comment (kaylareopelle) [inline metrics_sdk/test/opentelemetry/sdk/metrics/aggregation/exponential_histogram/exponential_bucket_histogram_test.rb:866]: I don't understand this comment. Isn't this the file we're in?

File: metrics_sdk/test/opentelemetry/sdk/metrics/aggregation/exponential_histogram/exponential_bucket_histogram_test.rb [hunk 6]
```diff
@@ -0,0 +1,1052 @@
...
+
+        results[1..].each_with_index do |metrics_data, _index|
+          metric_data = metrics_data[0]
+
+          _(metric_data.start_time_unix_nano).must_equal(previous_metric_data.start_time_unix_nano)
+          _(metric_data.min).must_equal(previous_metric_data.min)
+          _(metric_data.max).must_equal(previous_metric_data.max)
+          # _(metric_data.sum).must_be_within_epsilon(previous_metric_data.sum, 1e-10)
+
+          # omit the test case for now before cumulative aggregation is tested
```
Comment (kaylareopelle) [inline metrics_sdk/test/opentelemetry/sdk/metrics/aggregation/exponential_histogram/exponential_bucket_histogram_test.rb]: Do we still need to omit this case since the PR is fixing the merge logic for cumulative?

File: metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb [hunk 7]
```diff
@@ -217,11 +347,77 @@ def validate_scale(scale)
...
           def validate_size(size)
-            raise ArgumentError, "Max size #{size} is smaller than minimum size #{MIN_MAX_SIZE}" if size < MIN_MAX_SIZE
-            raise ArgumentError, "Max size #{size} is larger than maximum size #{MAX_MAX_SIZE}" if size > MAX_MAX_SIZE
+            raise ArgumentError, "Buckets min size #{size} is smaller than minimum min size #{MIN_MAX_SIZE}" if size < MIN_MAX_SIZE
+            raise ArgumentError, "Buckets max size #{size} is larger than maximum max size #{MAX_MAX_SIZE}" if size > MAX_MAX_SIZE
 
             size
           end
+
+          # checked, only issue is if @previous_scale is nil, then get_low_high may throw error
```
Comment (kaylareopelle) [inline metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb:356]: Would this error stop a user's application? Or is it rescued somewhere?

File: metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb [hunk 8]
```diff
@@ -217,11 +347,77 @@ def validate_scale(scale)
...
+            if current_low > current_high
+              [previous_low, previous_high]
+            elsif previous_low > previous_high
+              [current_low, current_high]
+            else
+              [[previous_low, current_low].min, [previous_high, current_high].max]
+            end
+          end
+
+          # checked
```
Comment (kaylareopelle) [inline metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb]: Do we still need this?

File: metrics_sdk/test/opentelemetry/sdk/metrics/aggregation/exponential_histogram/exponential_bucket_histogram_test.rb (hunk 5 above)
Comment (xuan-cao-swi) [reply metrics_sdk/test/opentelemetry/sdk/metrics/aggregation/exponential_histogram/exponential_bucket_histogram_test.rb:866]: Sorry, I was meant to move [python integration test](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/tests/metrics/integration_test/test_exponential_bucket_histogram.py) into this file. Python exp histogram separates the integration test out of main exp histogram test case.

File: metrics_sdk/test/opentelemetry/sdk/metrics/aggregation/exponential_histogram/exponential_bucket_histogram_test.rb (hunk 6 above)
Comment (xuan-cao-swi) [reply metrics_sdk/test/opentelemetry/sdk/metrics/aggregation/exponential_histogram/exponential_bucket_histogram_test.rb]: No, the comment is outdated (removed)

File: metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_histogram/exponent_mapping.rb (hunk 3 above)
Comment (xuan-cao-swi) [reply metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_histogram/exponent_mapping.rb]: It was used for scale check, but now it's not used since we check scale at exponential histogram class

File: metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb (hunk 1 above)
Comment (xuan-cao-swi) [reply metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb]: Removed

File: metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb (hunk 2 above)
Comment (xuan-cao-swi) [reply metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb]: Removed

File: metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb (hunk 7 above)
Comment (xuan-cao-swi) [reply metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb:356]: Sorry for the confusion. `@previous_scale[attributes] = current_scale if @previous_scale[attributes].nil?` will ensure `@previous_scale/previous_scale` never be nil.

File: metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb (hunk 8 above)
Comment (xuan-cao-swi) [reply metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_bucket_histogram.rb]: Removed

File: metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_histogram/logarithm_mapping.rb (hunk 4 above)
Comment (xuan-cao-swi) [reply metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/exponential_histogram/logarithm_mapping.rb]: Removed

**Review by kaylareopelle — APPROVED** [review]
